experiments/state/src/main/scala/org/apache/gearpump/streaming/state/system/impl/WindowState.scala

```diff
@@ -0,0 +1,163 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding cstateyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a cstatey of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.gearpump.streaming.state.system.impl
+
+import akka.actor.ActorSystem
+import com.twitter.algebird.Group
+import org.apache.gearpump._
+import org.apache.gearpump.cluster.UserConfig
+import org.apache.gearpump.streaming.state.system.api.MonoidState
+import org.apache.gearpump.streaming.task.TaskContext
+import org.apache.gearpump.util.LogUtil
+import org.slf4j.Logger
+
+import scala.collection.immutable.TreeMap
+
+/**
+ * an interval is a dynamic time range that is divided by window boundary and checkpoint time
+ */
+case class Interval(startTime: TimeStamp, endTime: TimeStamp) extends Ordered[Interval] {
+  override def compare(that: Interval): Int = {
+    if (startTime < that.startTime) -1
+    else if (startTime > that.startTime) 1
+    else 0
+  }
+}
+
+object WindowState {
+  val LOG: Logger = LogUtil.getLogger(classOf[WindowState[_]])
+
+  def apply[T](taskContext: TaskContext, conf: UserConfig)(implicit group: Group[T], system: ActorSystem): WindowState[T] = {
+    val windowConfig = conf.getValue[WindowConfig](WindowConfig.NAME).get
+    val windowManager = new WindowManager(windowConfig)
+    new WindowState(group, taskContext, windowManager)
+  }
+}
+
+/**
+ * this is a list of states, each of which is bounded by a time window
+ * state of each window doesn't affect each other
+ *
+ * WindowState requires a Algebird Group to be passed in
+ * Group augments Monoid with a minus function which makes it
+ * possible to undo the update by messages that have left the window
+ */
+class WindowState[T](group: Group[T],
+                     taskContext: TaskContext,
+                     windowManager: WindowManager)
```

Comment:
It will make sense if rename windowManager to window.

Then the window concept perfectly match WindowState.